docs/design/storagenode-graceful-exit.md

```diff
@@ -91,6 +92,134 @@ This process including the Storage nodes transferring their pieces to other node
 	- Receiving storage nodes should be uploaded to as if they were receiving any other upload (orders, etc).
 - We need to have the storage node keep track how its progress during graceful exit by calculating how much data it had when it started the graceful exit and keeping track of how much data it has deleted (data is deleted from the exiting node as it sends pieces to new nodes). This is an estimation since the storage node is likely holding more data than it will gracefully exit (garbage data). 
 
+## Implementation (MVP)
+
+#### Satellite
+- Update DBX - Add updateable `exit_initiated_at` and `exit_completed_at` timestamps to nodes table with indexes
+  - Add GetExitingNodeIds method to overlaycache. Returns node IDs where `exit_initiated_at` is not null and `exit_completed_at` is null.
+  - Add GetExitedNodeIds method to overlaycache. Returns node IDs where `exit_initiated_at` is not null and `exit_completed_at` is not null. 
+- Create GracefulExit endpoint
+  - Endpoints should be secured using the peer Identity provided in context
+  - Initiates the exit by setting `nodes.exit_initiated` to current time
+  - ``` 
+	service GracefulExit {
+		rpc InitiateExit(InitiateExitRequest) returns (InitiateExitResponse) {}
+	}
+
+	message InitiateExitRequest {
+	}
+
+	message InitiateExitResponse {
+	}
+	```
+- Update DBX - Add table exit_pieceinfos
+-  ```
+	model exit_pieceinfo (
+		key node_id path
+
+		field node_id           blob
+		field path              blob
+		field piece_info        blob
+		field durability_ratio  float64
+		field queued_at         timestamp ( autoinsert )
+		field sent_at           timestamp ( updateable )
+		field completed_at      timestamp ( updateable )
+	)
+   ```
+- Add `PieceAction` field to, `cache.FindStorageNodesRequest`. Update `cache.FindStorageNodesWithPreferences` to ignore exiting nodes for uploads and repairs.
+- Update Repairer service
+  - Modify `checker` to check segments for pieces associated with a storage node that is exiting. Add to `exit_pieceinfo` table if matches criteria.
```

Comment:
This should be implemented as a new observer to the metainfo loop.

Reply:
updated